feat(aws provider): Add support for `assume_role` to all AWS sinks

bruceg opened the pull request:
Closes #1556

Commit:
Create generic `create_client` implementation

Extracted from aws_cloudwatch_logs implementation.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add support for `assume_role` to `aws_kinesis_firehose` sink

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add `assume_role` support to `aws_cloudwatch_metrics` sink

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Finish `assume_role` for last sinks

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
Note: I'm not entirely happy with the `aws_s3` `create_client` function. Should that test access key logic be hoisted up into the shared `rusoto::create_client` function?

binarylogic reviewed (approved):
Docs look great

binarylogic committed:
Merge branch 'master' into aws-assume-role

Commit:
Drop unused macro pattern

Signed-off-by: Bruce Guenter <bruce@timber.io>

LucioFranco commented on src/sinks/util/rusoto.rs:
```diff
@@ -17,3 +26,60 @@ pub fn client(resolver: Resolver) -> crate::Result<Client> {
+        P::Future: Send;
+}
+
+pub fn create_client<T: RusotoNewClient>(
```
I find this trait a bit overkill for this imo. I think it'd be better to let each sink construct their concrete type via `T::new_with` within the sink, then have one fn to return the client and one fn to return a credentials provider. Then from there we can do something like

```rust
pub enum AwsCredentialsProvider {
	Default(DefaultCredentialsProvider),
	AssumeRole(...),
	Static(...),
}

fn client(...) -> crate::Result<Client> { ... }
fn credentials(...) -> crate::Result<AwsCredentialsProvider> { ... }
```

Imo this would be a bit easier to grok in a per sink sense and would avoid a set of macros and a trait.

bruceg replied:
It took a little more work this way, but the approach makes sense and is more composable. I'll see about adding the static provider to avoid that block in the S3 sink.

LucioFranco commented on src/sinks/util/rusoto.rs:
```diff
@@ -4,11 +4,20 @@ use hyper_openssl::{
-pub fn client(resolver: Resolver) -> crate::Result<Client> {
+#[derive(Debug, Snafu)]
+enum RusotoError {
+    #[snafu(display("{}", source))]
```
```suggestion
    #[snafu(display("Invalid AWS credentials: {}", source))]
```

Do we want to add some sort of context here?

bruceg replied:
That was cloned from the `aws_cloudwatch_logs` sink. I agree it should have more context. I'll add it.

LucioFranco reviewed (commented):
Overall looks good, I left you a proposal to possible make this a bit easier to read in the future. Happy to go either way :)

Commit:
Improve invalid credentials error message

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add a new AWS credentials provider

This allows us to wrap the different kinds of credentials we provide,
making the resulting code more composable, and eliminates the new client
trait and split implementations.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into aws-assume-role

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move the support for minimal static AWS credentials to the lib

Signed-off-by: Bruce Guenter <bruce@timber.io>

LucioFranco commented on src/sinks/util/rusoto.rs:
```diff
@@ -49,6 +54,13 @@ impl AwsCredentialsProvider {
         }
     }
+
+    pub fn new_minimal<A: Into<String>, S: Into<String>>(access_key: A, secret_key: S) -> Self {
```
minor nit: I would probably use `impl Trait` here, I find it much nicer to the eyes.

LucioFranco commented on src/sinks/util/rusoto.rs:
```diff
@@ -1,13 +1,97 @@
+        }
+    }
+
+    pub fn new_minimal<A: Into<String>, S: Into<String>>(access_key: A, secret_key: S) -> Self {
```
Also curious if this should be called `new_static`?

bruceg replied:
I thought about that, but the static provider has both `new_minimal` and a plain `new`, and I thought I'd reflect which one is being called, giving room for a `new_static` to provide the full parameters if needed.

LucioFranco replied:
Ah ok that is fine!

LucioFranco reviewed (approved):
LGTM! Left some non blocking clients. :ship: it
